src/migrate.c

```diff
@@ -0,0 +1,166 @@
+        ShardGroupSlotRange *sr = &sg->slot_ranges[i];
+        if (sr->start_slot <= slot && sr->end_slot >= slot && sr->magic == magic) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+static int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    int slot = KeyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        // this is acceptable (suspended old leader
+        LOG_WARNING("ignoring import keys as old term");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid term");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMagic(rr, slot, import_keys.magic)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid magic");
+        }
+        goto exit;
+    }
+
+    RedisModuleString * zero = RedisModule_CreateString(rr->ctx, "0", 1);
+
+    for (size_t i = 0; i < import_keys.num_keys; i++) {
+        RedisModuleString * temp[3];
+        temp[0] = import_keys.key_names[i];
+        temp[1] = zero;
+        temp[2] = import_keys.key_serialized[i];
+
+        if (!RedisModule_KeyExists(rr->ctx, temp[0])) {
+            enterRedisModuleCall();
+            RedisModuleCallReply *reply;
+            RedisModule_Assert((reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 3)) != NULL);
+            exitRedisModuleCall();
+            RedisModule_Assert(RedisModule_CallReplyType(reply) != REDISMODULE_REPLY_ERROR);
+            RedisModule_FreeCallReply(reply);
+        }
+    }
+
+    RedisModule_FreeString(rr->ctx, zero);
+
+    if (req) {
+        RedisModule_ReplyWithSimpleString(req->ctx, "OK");
+    }
+
+    exit:
+    if (req) {
+        RaftReqFree(req);
+    }
+    FreeImportKeys(&import_keys);
+}
+
+int cmdRaftImport(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
+{
+    RedisRaftCtx *rr = &redis_raft;
+
+    if (argc < 3 || ((argc -3)) % 2 != 0) {
```

Comment:
argc < 5 not 3.  add comment that 5 implies the 3 strings raft.import, magic, term and the next 2 are the first key/serialized combo